### PR TITLE
C_LAPACK: Fixes to make it compile with MSVC

### PR DESCRIPTION
**Pull request by martin-frbg**

### CODE REVIEW

File: lapack-netlib/INSTALL/dlamch.c [hunk 1]
```diff
@@ -267,6 +291,21 @@ static double dpow_ui(double x, integer n) {
...
 	return pow;
 }
+#ifdef _MSC_VER
+static _Fcomplex cpow_ui(complex x, integer n) {
+	complex pow={1.0,0.0}; unsigned long int u;
+		if(n != 0) {
+		if(n < 0) n = -n, x.r = 1/x.r, x.i=1/x.i;
+		for(u = n; ; ) {
+			if(u & 01) pow.r *= x.r, pow.i *= x.i;
+			if(u >>= 1) x.r *= x.r, x.i *= x.i;
```
Comment (isuruf) [inline lapack-netlib/INSTALL/dlamch.c:301]: This seems to assume that `(a + b*I)^n = a^n + b^n*I`

Comment (martin-frbg) [reply lapack-netlib/INSTALL/dlamch.c:301]: Yes, I admit to making some scary assumptions about complex arithmetic here - math teacher would not be proud. Been meaning to look this up if anything actually uses it (a handful of files do) - provided it makes sense to pursue this. (Pure MSVC builds are guaranteed low performance as they cannot use our assembly kernels, so perhaps should not be encouraged)

**Comment by mmuetzel** [comment]
Much of these changes seem to repeat themselves many times. Given that this is already deviating quite significantly from the f2c translated sources, would it make sense to move these functions into a common header (or multiple headers if that makes more sense) that could be included in those files?

**Comment by martin-frbg** [comment]
Yes this is extremely repetitive as I had decided against using an extra header (naively assuming this to be a one-time edit...)

File: lapack-netlib/INSTALL/dlamch.c [hunk 2]
```diff
@@ -228,11 +252,11 @@ static char junk[] = "\n@(#)LIBF77 VERSION 19990503\n";
...
 #define myexit_() break;
-#define mycycle_() continue;
-#define myceiling_(w) ceil(w)
-#define myhuge_(w) HUGE_VAL
+#define mycycle() continue;
+#define myceiling(w) {ceil(w)}
+#define myhuge(w) {HUGE_VAL}
 //#define mymaxloc_(w,s,e,n) {if (sizeof(*(w)) == sizeof(double)) dmaxloc_((w),*(s),*(e),n); else dmaxloc_((w),*(s),*(e),n);}
-#define mymaxloc_(w,s,e,n) dmaxloc_(w,*(s),*(e),n)
+#define mymaxloc(w,s,e,n) {dmaxloc_(w,*(s),*(e),n)}
```
Comment (mmuetzel) [inline lapack-netlib/INSTALL/dlamch.c:259]: Should this be the following (untested)?
```
#ifdef _MSC_VER
#define myceiling_(w) {ceil(w)}
#define myhuge_(w) {HUGE_VAL}
#define mymaxloc_(w,s,e,n) {dmaxloc_(w,*(s),*(e),n)}
#else
#define myceiling_(w) ceil(w)
#define myhuge_(w) HUGE_VAL
#define mymaxloc_(w,s,e,n) dmaxloc_(w,*(s),*(e),n)
#endif
```

File: lapack-netlib/INSTALL/dlamch.c [hunk 3]
```diff
@@ -170,8 +189,13 @@ typedef struct Namelist Namelist;
 #define abort_() { sig_die("Fortran abort routine called", 1); }
 #define c_abs(z) (cabsf(Cf(z)))
 #define c_cos(R,Z) { pCf(R)=ccos(Cf(Z)); }
+#ifdef _MSC_VER
+#define c_div(c, a, b) {Cf(c)._Val[0] = (Cf(a)._Val[0]/Cf(b)._Val[0]); Cf(c)._Val[1]=(Cf(a)._Val[1]/Cf(b)._Val[1]);}
```
Comment (mmuetzel) [inline lapack-netlib/INSTALL/dlamch.c:193]: Should probably be (untested):
```
#define c_div(c, a, b) { Cf(c)._Val[0] = (Cf(a)._Val[0]*Cf(b)._Val[0]-Cf(a)._Val[1]*Cf(b)._Val[1])/(Cf(b)._Val[0]*Cf(b)._Val[0]-Cf(b)._Val[1]*Cf(b)._Val[1]); Cf(c)._Val[1] = (Cf(a)._Val[1]*Cf(b)._Val[0]-Cf(b)._Val[1]*Cf(a)._Val[0])/(Cf(b)._Val[0]*Cf(b)._Val[0]-Cf(b)._Val[1]*Cf(b)._Val[1]); }
```
And similarly for `z_div`.

File: lapack-netlib/INSTALL/dlamch.c (hunk 1 above)
Comment (mmuetzel) [reply lapack-netlib/INSTALL/dlamch.c:301]: You could probably use the `cpow` family of functions for the MSVC implementation:
https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/cpow-cpowf-cpowl?view=msvc-170

The above alternative implementation isn't correct afaict.

File: lapack-netlib/INSTALL/dlamch.c [hunk 4]
```diff
@@ -324,6 +380,22 @@ static integer smaxloc_(float *w, integer s, integer e, integer *n)
 }
 static inline void cdotc_(complex *z, integer *n_, complex *x, integer *incx_, complex *y, integer *incy_) {
 	integer n = *n_, incx = *incx_, incy = *incy_, i;
+#ifdef _MSC_VER
+	_Fcomplex zdotc = {0.0, 0.0};
+	if (incx == 1 && incy == 1) {
+		for (i=0;i<n;i++) { /* zdotc = zdotc + dconjg(x(i))* y(i) */
+			zdotc._Val[0] += conjf(Cf(&x[i]))._Val[0] * Cf(&y[i])._Val[0];
+			zdotc._Val[1] += conjf(Cf(&x[i]))._Val[1] * Cf(&y[i])._Val[1];
```
Comment (mmuetzel) [inline lapack-netlib/INSTALL/dlamch.c:388]: Should this be the following (untested)?
```
zdotc._Val[0] += Cf(&x[i])._Val[0]*Cf(&y[i])._Val[0] + Cf(&x[i])._Val[1]*Cf(&y[i])._Val[1];
zdotc._Val[1] += Cf(&x[i])._Val[0]*Cf(&y[i])._Val[1] - Cf(&x[i])._Val[1]*Cf(&y[i])._Val[0];
```

And similarly below.

**Comment by mmuetzel** [comment]
> Yes this is extremely repetitive as I had decided against using an extra header (naively assuming this to be a one-time edit...)

It's not too late to move this to a header now. 🙂

**Pull request merged**